src/chainstate/stacks/db/blocks.rs

```diff
@@ -4996,6 +5003,25 @@ impl StacksChainState {
         query_row(&self.db(), sql, args).map_err(Error::DBError)
     }
 
+    /// This runs checks for the validity of a transaction that
+    ///   can be performed just by inspecting the transaction itself (i.e., without
+    ///   consulting chain state).
+    fn can_admit_mempool_semantic(tx: &StacksTransaction) -> Result<(), MemPoolRejection> {
+        match tx.payload {
+            TransactionPayload::TokenTransfer(ref recipient, amount, ref _memo) => {
+                let origin = PrincipalData::from(tx.origin_address());
+                if &origin == recipient {
+                    return Err(MemPoolRejection::TransferRecipientIsSender(origin));
+                }
+                if amount == 0 {
+                    return Err(MemPoolRejection::TransferAmountMustBePositive);
+                }
```

Comment:
Could we also check that testnet addresses are used on testnets, and mainnet addresses on mainnet?

Reply:
Yes, will add